h2o-r/tests/testdir_algos/glm/runit_GLM_negbinomial_dispersion.R

```diff
@@ -0,0 +1,259 @@
+
+    hdata <- as.h2o(data)
+    hglm <- tryCatch(h2o.glm(y = "result", training_frame = hdata, family = "negativebinomial", link = link, dispersion_parameter_method = "ml", standardize = FALSE),
+                     error = function(e) NULL)
+    expect_true(!is.null(hglm))
+    if (is.null(hglm)) {
+        print(summary(data))
+        return()
+    }
+
+    hdiff <- abs(hglm@model$coefficients_table$coefficients - c(intercept, coefs))
+    hdiff_mean <- mean(hdiff)
+
+    hloglik <- get_h2o_likelihood(hglm, hdata, data)
+    rloglik <- logLik(rglm)
+
+    hMSE <- h2o.performance(hglm, as.h2o(valid_data))@metrics$MSE
+    rMSE <- mean((valid_data$result-predict(rglm, valid_data, type = "response"))**2)
+
+    hMAE <- h2o.performance(hglm, as.h2o(valid_data))@metrics$mae
+    rMAE <- mean(abs(valid_data$result-predict(rglm, valid_data, type = "response")))
+
+    mse_test <- hMSE <= tolerance * rMSE
+    mae_test <- hMAE <= tolerance * rMAE
+    likelihood_test <- hloglik >=  rloglik - abs((tolerance - 1) * rloglik)
+    #dispersion_test <- abs(hglm@params$actual$theta - dispersion) <= tolerance * abs(dispersion - 1/rglm$theta)
+    dispersion_test <- abs(hglm@params$actual$theta - 1/rglm$theta) <= (tolerance - 1)*dispersion ||  # Either the estimate is close to R's
+                       abs(hglm@params$actual$theta - dispersion) <= tolerance * abs(dispersion - 1/rglm$theta) # Or it's closer to the true value than R
+    coef_test <- hdiff_mean <= tolerance * rdiff_mean
+
+    if (!dispersion_test) {
+        h2o.no_progress({
+            cat(msg, "\n")
+            cat("Likelihood test: ", likelihood_test, "; MSE test: ", mse_test, "; MAE test: ", mae_test, "; Dispersion test: ", dispersion_test, "; Coefficient test: ", coef_test, "\n")
+            cat(" r loglikelihood: ", logLik(rglm), "; h2o loglikelihood: ", get_h2o_likelihood(hglm, hdata, data), "\n",
+                "r 1/theta: ", 1/rglm$theta, "; h2o theta: ", hglm@params$actual$theta, "; true dispersion: ", dispersion, "\n",
+                "mean(abs(r - true)) coef diff: ", rdiff_mean, "; mean(abs(h2o - true)) coef diff: ", hdiff_mean, "; mean(abs(r - h2o)) coef diff: ",
+                 mean(abs(hglm@model$coefficients_table$coefficients - coefficients(rglm))), "\n",
+                "r validation MSE: ", rMSE, "; h2o validation MSE: ", hMSE, "\n",
+                "r validation MAE: ", rMAE, "; h2o validation MAE: ", hMAE, "\n")
+            cat("\nTrue Coefficients:\n")
+            print(stats::setNames(c(intercept, coefs), c("(Intercept)", paste0("col", seq_along(coefs)-1))))
+            cat("\nH2O Coefficients:\n")
+            print(hglm@model$coefficients_table)
+            cat("\nR Coefficients:\n")
+            print(coefficients(rglm))
+            cat("\nResponse summary:\n")
+            print(summary(data$result))
+        })
+    }
+
+    expect_true(likelihood_test)
+    expect_true(mse_test)
+    expect_true(dispersion_test)
+    expect_true(coef_test)
```

Comment:
This will change to just `expect_true(dispersion_test)` but for developing it's nice to see the impact of the changes. For some cases I'm unable to get "better" dispersion than R (closer to the true) but at the same time I get better likelihood so maybe in those cases I get closer to the "true" sample dispersion.

To monitor the changes I use the following quick and dirty AWK script on R's stdout:
```awk
#! /usr/bin/awk -f

BEGIN {
    activeRun="none"
}
/Running .*/ {
    activeRun=$2
}

/Dispersion: [[:digit:]]+.?[[:digit:]]* ; seed: [[:digit:]]+/ {
    numOfTests += 1
}

/dropCols/ {
    failureDropCols[activeRun] += 1
}

/NonSPDMatrixException/ {
    failureSPD[activeRun] += 1
}

/no valid set of coefficients has been found: please supply starting values/ {
    validCoefError[activeRun] += 1
}

/R model training failed/ {
    failure[activeRun] += 1
}
/Reached maximum number of iterations 50/ {
    maxIter[activeRun] += 1
}

/Likelihood test:  FALSE/ {
    llhFails[activeRun] += 1
}

/Likelihood test:  TRUE/ {
    llhSucc[activeRun] += 1
}

/Dispersion test:  FALSE/ {
    dispFail[activeRun] += 1
}

/MSE test:  FALSE/ {
    mseFail[activeRun] += 1
}

/MAE test:  FALSE/ {
    maeFail[activeRun] += 1
}

/Coefficient test:  FALSE/ {
    coefFail[activeRun] += 1
}

/NA\/NaN\/Inf in 'x'/ {
    naInfFail[activeRun] += 1
}

END {
    OFS=": "

    metricFailures = 0
    print "\nLikelihood failures:"
    total = 0
    for (i in llhFails) {
        print i, llhFails[i]
        total += llhFails[i]
        metricFailures += llhFails[i]
    }
    print "Total", total

    # print "\nLikelihood successes:"
    # for (i in llhSucc)
    #     print i, llhSucc[i]

    print "\nMSE failures:"
    total = 0
    for (i in mseFail){
        print i, mseFail[i]
        total += mseFail[i]
        metricFailures += mseFail[i]
    }
    print "Total", total

    print "\nMAE failures:"
    total = 0
    for (i in maeFail) {
        print i, maeFail[i]
        total += maeFail[i]
        metricFailures += maeFail[i]
    }
    print "Total", total

    print "\nDispersion failures:"
    total = 0
    for (i in dispFail) {
        print i, dispFail[i]
        total += dispFail[i]
        metricFailures += dispFail[i]
    }
    print "Total", total
    dispFailuresTotal = total

    print "\nCoef failures:"
    total = 0
    for (i in coefFail){
        print i, coefFail[i]
        total += coefFail[i]
        metricFailures += coefFail[i]
    }
    print "Total", total

    print "\nDropCols:"
    total = 0
    for (i in failureDropCols) {
        print i, failureDropCols[i]
        total += failureDropCols[i]
    }
    print "Total", total

    print "\nNonSPDMatrixException:"
    total = 0
    for (i in failureSPD) {
        print i, failureSPD[i]/2
        total += failureSPD[i]/2
    }
    print "Total", total

    print "\nModel Failures:"
    total = 0
    for (i in failure) {
        print i, failure[i]
        total += failure[i]
    }
    print "Total", total
    Rfails = total

    print "\nReached MaxIter:"
    total = 0
    for (i in maxIter) {
        print i, maxIter[i]
        total += maxIter[i]
    }
    print "Total", total

    print "\nValid set of coefs not found (R error):"
    total = 0
    for (i in validCoefError) {
        print i, validCoefError[i]
        total += validCoefError[i]
    }
    print "Total", total

    print "\nError in glm.fitter: NA/NaN/Inf in 'x' (R error):"
    total = 0
    for (i in naInfFail) {
        print i, naInfFail[i]
        total += naInfFail[i]
    }
    print "Total", total

    print "\n\nNumber of tests ran", numOfTests
    print "Number of successful tests", numOfTests - Rfails - dispFailuresTotal
    print "Failed tests", metricFailures
    print "Dispersion failures", dispFailuresTotal
}

```